### ISSUE TITLE
Relay IDs and scalar types should be opaque types in flow

### ISSUE THREAD

**Issue by sheki** [facebook/relay#2567]
Currently the flow types generated by relay use `string` type to represent the ID. The type for the ID field should be a flow opaque type so that users on cannot pass it arbitrary strings as IDs.

This behaviour should be extended for custom scalar types. I.e. Every scalar type should be a unique opaque type. 

```
export type RestrictedPartyHitsPanelRestrictedPartyHitScanForCompanyMutationResponse = {|
  +RestrictedPartyHitScanForCompany: ?{|
    +company: ?{|
      +id: string,  // CAN THIS BE A an opaque type
....
```

**Comment by AugustinLF** [comment]
It makes sense, but in my use project, we serialize/deserialize ids, so a user can navigate to `user/2` and fetch the user of id 2, with relay. We use on the client and the server the same logic for this serialization problem. Wouldn't using opaque types a problem here? I'm all for saying that we have to pass through a serialization function to generate the opaque type (since we don't want people to do that on the fly), but if I recall correctly, opaque types prevent that, right?

**Comment by josephsavona** [comment]
@AugustinLF Note that Flow opaque types allow opting in to subtyping. I suspect we'd end up with something like `opaque type ID: string = string;`, where means that you can't use an arbitrary string as an ID, but you can pass an ID to a function that expects a string.

**Comment by AugustinLF** [comment]
Hmmm, it works for deserializing then, but how I am suppose to do `(userId: string) -> relayId`? Because that's part of my use-case.

**Comment by AugustinLF** [comment]
Yeah, sure, it's always a possibility. Just sad that if my use-case is considered "valid" (might be a bad practice, I don't know), the typings shouldn't work against me, right?

**Comment by josephsavona** [comment]
> we’d be willing to accept a PR that allowed you to specify to the compiler a map of scalar types to modules/types to import and use

Yeah that seems reasonable.

**Comment by sheki** [comment]
Here is a tiny PR to get early feedback https://github.com/facebook/relay/pull/2571
